doc/tutorials/optimizer/optimizer.txt

```diff
@@ -0,0 +1,251 @@
+implement. \c mlpack offers a few variations of this API to cover different
+function characteristics.
+
+Any \c FunctionType requires the implementation of an \c Evaluate() method.
+The interface used for that can be one of the following two methods:
+
+@code
+// For non-separable objectives.
+double Evaluate(const arma::mat& parameters);
+
+// For separable objectives.
+double Evaluate(const arma::mat& parameters,
+                const size_t start,
+                const size_t batchSize);
+@endcode
+
+Both of these methods calculate the objective given the parameters matrix \c
+parameters. The second overload is for separable functions, and should calculate
+the partial objective starting at the separable function indexed by start and
+calculate \c batchSize partial objectives and return the sum. Functions
+implementing the first overload are used by optimizers like the \ref
+GradientDescent optimizer; functions implementing the second are used by SGD-
+like optimizers.
+
+In addition, any differentiable function must implement some \c Gradient()
+method.
+
+@code
+// For non-separable differentiable sparse and non-sparse functions.
+template<typename GradType>
+void Gradient(const arma::mat& parameters, GradType& gradient);
+
+// For separable differentiable sparse and non-sparse functions.
+template<typename GradType>
+void Gradient(const arma::mat& parameters,
+              const size_t start,
+              GradType& gradient,
+              const size_t batchSize);
+@endcode
+
+Both of these methods should calculate the gradient and place the results into
+the matrix object \c gradient that is passed as an argument. Note that the
+method accepts a template parameter \c GradType, which may be \c arma::mat
+(dense Armadillo matrix) or \c arma::sp_mat (sparse Armadillo matrix).  This
+allows support for both sparse-supporting and non-sparse-supporting optimizers.
+
+If the objective function is partially differentiable, we could implement the
+following method:
+
+@code
+// For partially differentiable sparse and non-sparse functions.
+template<typename GradType>
+void PartialGradient(const arma::mat& parameters, const size_t j, GradType& gradient);
+@endcode
+
+The function should calculate the gradient of the parameters \c parameters with
+respect to the parameter \c j and store the results (either sparse or dense) in
+the \c gradient matrix object.
+
+In addition, a separable partially differentiable functions must implement
+the \c NumFunctions() and or \c NumFeatures() functions, respectively:
+
+@code
+// For separable functions: return the number of parts the optimization problem
+// can be decomposed into.
+size_t NumFunctions();
+
+// For partially differentiable functions: return the number of partial derivatives.
+size_t NumFeatures();
+@endcode
+
+Finally, \c separable functions must implement the method \c Shuffle() function,
+which shuffles the ordering of the functions
```

Comment:
End of sentence punctuation (sorry if I'm being excessively concerned about grammar).